networkx/algorithms/centrality/knotty.py

```diff
@@ -0,0 +1,165 @@
+    Examples
+    --------
+    >>> G = nx.Graph([(0,1),(0,2),(1,2),(1,3),(1,4),(4,5)])
+    >>> nodes, kc = nx.knotty_centrality(G,compact=False)
+
+    Notes
+    ------
+    The knotty centrality definition and algorithm are found in [1]_. 
+
+    References
+    ----------
+    .. [1] Shanahan M, Wildie M (2012) Knotty-Centrality: 
+    Finding the Connective Core of a Complex Network. 
+    PLoS ONE 7(5): e36579. doi:10.1371/journal.pone.0036579 
+
+    """
+    if G.is_multigraph() or G.is_directed():
+        raise Exception('knotty centrality is not implemented for ',
+                        'directed or multiedge graphs.')
+    return _find_knotty_centre(G, compact)
+
+
+def _find_knotty_centre(G, compact):
+    # nodes = the sub-graph found
+    # kc = its knotty-centredness
+
+    N = G.number_of_nodes()
+
+    # binarise matrix - all non-zero weights become 1s
+    edge_matrix = np.array(nx.to_numpy_matrix(G))
+    CIJ = edge_matrix!=0
+    CIJ = CIJ.astype(int)
+
+
+    # Exhastive search phase
+    Exh = 5 # number of nodes for exhaustive search (2^Exh combinations)
+    Exh = min([Exh,N])
+    
+    BC_dict = nx.betweenness_centrality(G)
+    BC_list = list(BC_dict.values())
+    
+    BC = np.array(BC_list)/sum(BC_list) # normalise wrt total betweenness centrality
+
+    # Calculate indirect betweenness centrality
+    BC2 = BC + np.sum(CIJ*BC,axis=0) + np.sum(CIJ,axis=1).T*BC
+
+    # Ranking the nodes in descending order
+    array_asc = (BC2.ravel().argsort())
+    list_asc = array_asc.tolist()
+    list_asc.reverse() #list_asc is not descending
+    IxBC = list_asc
+    
+    nodes = []
+    improving = 1
+    
+    while improving:
+        L = len(nodes)
+        nodes_left = IxBC
+        nodes_left = [x for x in nodes_left if x not in set(nodes)]
+        choices = [nodes_left[i] for i in range(0, min([Exh,len(nodes_left)]))]
+        nodes, kc = _best_perm(nodes, choices, G, CIJ, compact, BC)
+        improving = len(nodes) > L
+
+    # Hill climbing phase
+    nodes_left = range(0,N)
+    nodes_left = [x for x in nodes_left if x not in set(nodes)]
+    improving = 1
+    
+    while improving and nodes_left:
+        best_kc = 0
+        for node in nodes_left:
+            nodes2 = np.hstack((nodes, node))
+            kc2 = _compute_knotty_centrality(G, CIJ, nodes2, compact, BC)
+            if kc2 > best_kc:
+                best_kc = kc2
+                best_node = node
+                
+        if best_kc > kc:
+            kc = best_kc
+            nodes = np.hstack((nodes, best_node))
+            nodes_left.remove(best_node)
+        else:
+            improving = 0
+    return nodes, kc
+
+def _best_perm(given, choices, G, CIJ, compact, BC):
+    # Carries out exhaustive search to find a permutation of nodes in
+    # "choices" that when added to the nodes in "given" yields the highest
+    # value of knotty-centrality
+
+    if choices:
+        choices2 = [choices[i] for i in range(0, len(choices))]
```

Comment:
Per the original paper, this should be

```python
choices2 = choices[1:]
```
I question the name of the function and its description.  It's clearly iterating over combinations, not permutations.

it would be more pythonic to write

```
def _best_perm(given, choices, G, CIJ, compact, BC):
    best_nodes = given
    best_kc = _compute_knotty_centrality(G, CIJ, given, compact, BC)
    for r in range(1, len(choices)+1):
        for subset in combinations(given, r):
            nodes = np.hstack((given, subset))
            kc = _compute_knotty_centrality(G, CIJ, nodes, compact, BC)
            if kc > best_kc:
                best_kc = kc
                best_nodes = nodes
    return best_nodes, best_kc
```

Reply:
Thanks @boothby!! That indeed fixes the recursion error.

Just to check the algorithm I did a naive implementation and the results don't seem to match up:
``` python
from itertools import chain, combinations


def knotty_centrality_naive(G):
    graph_subsets = _graph_powerset(set(G))
    bc = nx.betweenness_centrality(G)
    bc_sum = sum(bc.values())
    bc = {n: val/bc_sum for n, val in bc.items()}
    current_max = 0
    current_max_set = {}
    count = 0
    for subset in graph_subsets:
        kc_val = _knotty_cent(G, subset, bc)
        count += 1
        if kc_val > current_max:
            current_max = kc_val
            current_max_set = subset
            print(current_max, current_max_set)
    return current_max, current_max_set, count


def _graph_powerset(iterable):
    s = list(iterable)
    return chain.from_iterable(combinations(s, r) for r in range(2, len(s)+1))
    
def _knotty_cent(G, subset, bc):
    subset_graph = G.subgraph(subset)
    edges = subset_graph.number_of_edges()
    nodes = subset_graph.number_of_nodes()
    bc_subset = sum(bc[i] for i in subset)
    return (edges*bc_subset) / ((nodes - 1) * (nodes))

```

Gives the highest knotty_centrality of 0.333 but the implementation in the PR gives 0.666. 

And interestingly the paper states that knotty centrality is for directed graphs, @swederik do you remember by any chance what was the motivation to put in the error for a directed graph?

Reply:
In the paper, they define the knotty centrality for subgraphs which have at least two nodes, as you have implemented in your naive version.  It appears that the implementation here respects that condition.

Reply:
I think it was probably because I was not working with directed graphs at the time and did not test it. I ported over the MATLAB code after meeting the authors at the Brain Connectivity Workshop conference at the time. I don't recall if I ever ended up using it in an actual publication.